### ISSUE TITLE
Incorrect gradient for assignment

### ISSUE THREAD

**Issue by yongfei25** [twitter/torch-autograd#102]
Hi,

If you add a third assignment here, for example: `xc[3] = 1`, gradient check will return incorrect gradient.
I'm not sure is this issue with assignment or the gradient check.

https://github.com/twitter/torch-autograd/blob/86a5963dd6e3cfd9c3b29fcddcf2edb7c3759ac4/test/test.lua#L1471-L1472

**Comment by alexbw** [comment]
FYI, assignment doesn't work in non-optimized mode. This is pretty weird, @luketwitter any ideas?

**Comment by yongfei25** [comment]
Hey, would really like to get this fixed. Here's a bit of debugging I've tried.

```lua
local autograd = require 'autograd'
autograd.optimize(true)
torch.manualSeed(123)

local f2 = function(params)
   local xc = torch.clone(params.x)
   xc[1] = torch.sum(params.y)*2.0
   xc[2] = torch.sum(params.y)*3.0
   xc[3] = torch.sum(params.y)*4.0
   return torch.sum(xc)
end

local df = autograd(f2)
df({x=torch.randn(10),y=torch.randn(3)})

-- Generate the following codes
-- ...
rlocals[1][1] = ((torch_sum(p1.y)) * 2)
rlocals[1][2] = ((torch_sum(p1.y)) * 3)
locals[7] = torch_sum(rlocals[1])
rlocals[1][3] = ((torch_sum(p1.y)) * 4)
-- ...
```

It seem to be me that `rlocals[1][3]` is not taken into `torch_sum`.
I've also look into the internal of code compilation, do you any idea which area to start looking at?
